### ISSUE TITLE
uvicorn.run with reload or workers in ProcessPoolExecutor fails with Bad file descriptor error

### ISSUE THREAD

**Issue by likeanaxon** [Kludex/uvicorn#877]
Greetings!

I have run into an issue when trying to spawn several local `uvicorn`s (obviously bound to different ports) **from within Python**. Instead of launching each one manually in a separate terminal, I have set up a function that spawns the desired servers in a `ProcessPoolExecutor` (imported from `concurrent.futures`). However, if the servers are configured with the `workers` or `reload` options, I get the following error:

```
    Process SpawnProcess-1:1:
    Traceback (most recent call last):
      File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
        self.run()
      File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
        self._target(*self._args, **self._kwargs)
      File "/home/cantor/code/envs/tabula/lib/python3.8/site-packages/uvicorn/subprocess.py", line 55, in subprocess_started
        sys.stdin = os.fdopen(stdin_fileno)
      File "/usr/lib/python3.8/os.py", line 1023, in fdopen
        return io.open(fd, *args, **kwargs)
    OSError: [Errno 9] Bad file descriptor

```

Stating the obvious, there is an issue with reopening `stdin` in the child process.

Simple workarounds resulting in happy server spawning: 

- Commenting out lines `54` and `55` in `subprocess.py`

```
    # if stdin_fileno is not None:
        # sys.stdin = os.fdopen(stdin_fileno)
```

- Setting `'stdin_fileno': None` on line `35` in `subprocess.py`

### To reproduce

```
pip install fastapi uvicorn
```

(Assuming the code is located in `launch.py`)

With `workers`:

```
    import uvicorn
    from fastapi import FastAPI
    from concurrent.futures import ProcessPoolExecutor

    app = FastAPI()

    if __name__ == '__main__':

        with ProcessPoolExecutor() as executor:

            executor.submit(uvicorn.run, 'launch:app', workers = 3)
```

Error:

```
    Process SpawnProcess-1:3:
    Traceback (most recent call last):
    File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
        self.run()
    File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
        self._target(*self._args, **self._kwargs)
    File "/usr/lib/python3.8/site-packages/uvicorn/subprocess.py", line 55, in subprocess_started
        sys.stdin = os.fdopen(stdin_fileno)
    File "/usr/lib/python3.8/os.py", line 1023, in fdopen
        return io.open(fd, *args, **kwargs)
    OSError: [Errno 9] Bad file descriptor
    Process SpawnProcess-1:1:
    Traceback (most recent call last):
    File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
        self.run()
    File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
        self._target(*self._args, **self._kwargs)
    File "/usr/lib/python3.8/site-packages/uvicorn/subprocess.py", line 55, in subprocess_started
        sys.stdin = os.fdopen(stdin_fileno)
    File "/usr/lib/python3.8/os.py", line 1023, in fdopen
        return io.open(fd, *args, **kwargs)
    OSError: [Errno 9] Bad file descriptor
    Process SpawnProcess-1:2:
    Traceback (most recent call last):
    File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
        self.run()
    File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
        self._target(*self._args, **self._kwargs)
    File "/usr/lib/python3.8/site-packages/uvicorn/subprocess.py", line 55, in subprocess_started
        sys.stdin = os.fdopen(stdin_fileno)
    File "/usr/lib/python3.8/os.py", line 1023, in fdopen
        return io.open(fd, *args, **kwargs)
    OSError: [Errno 9] Bad file descriptor
```

With `reload`:

```
    import uvicorn
    from fastapi import FastAPI
    from concurrent.futures import ProcessPoolExecutor

    app = FastAPI()

    if __name__ == '__main__':

        with ProcessPoolExecutor() as executor:

            executor.submit(uvicorn.run, 'launch:app', reload = True)
```

Error:

```
    Process SpawnProcess-1:1:
    Traceback (most recent call last):
    File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
        self.run()
    File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
        self._target(*self._args, **self._kwargs)
    File "/usr/lib/python3.8/site-packages/uvicorn/subprocess.py", line 55, in subprocess_started
        sys.stdin = os.fdopen(stdin_fileno)
    File "/usr/lib/python3.8/os.py", line 1023, in fdopen
        return io.open(fd, *args, **kwargs)
    OSError: [Errno 9] Bad file descriptor
```

With `'stdin_fileno': None` on line `35` in `subprocess.py`:

```
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [705451] using statreload
INFO:     Started server process [705462]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
```

### Expected behavior

`uvicorn` server starting normally with a reloader or one or more workers.

### Actual behavior

Server runs, but no reloader or worker subprocesses.

### Environment

- OS: Ubuntu 20.04
- Python: CPython 3.8.6
- Uvicorn version: 0.12.3

**Comment by euri10** [comment]
Thanks for the detailed report,

I'm not sure what to say on this, I don't remember why this `sys.stdin = os.fdopen(stdin_fileno)` was added, some `git log -S` magic could help, 

That said, I'm not sure if we should support / encourage that kind of usage : both --worker and --reload flag spawn new processes, using the spawn context. I'm not familiar enough with Multiprocessing enough to tell you if using a pool to spawn processes that themselves will spawn processes is a good idea or not, it just feels awkward to me and kind of an edge case that you eventually could deal in a different manner. I'm not saying your use-case is not legit though.

I'd be tempted to tag it as wont fix, but maybe others will have a more constructive and positive approach about this, let's keep it open for now !

**Comment by likeanaxon** [comment]
@euri10 Thank you for the feedback!

I actually agree with your point about this setup being awkward. I originally had a slightly different implementation - running the servers in a `ThreadPoolExecutor` rather than a `ProcessPoolExecutor`. In that case, I hit a different issue (I cannot remember exactly, but I think it had to do with signal handling) which pushed me in the direction of using a process pool. I noticed [here](https://github.com/encode/uvicorn/releases/tag/0.13.0) that the installation of signal handlers will be skipped when not in the main thread, so I might give the `ThreadPoolExecutor` another go when 0.13.0 is officially released.

However, in the meantime...

Given that there is potentially no real reason for `sys.stdin` to be reopened in the child processes, and given that the cause of the problem is very clear, would it be at least possible to put the offending code (lines 54 and 55 in `subprocess.py`) in a `try` / `except` block so that it doesn't crash the child process? Awkwardness aside, my setup actually works really well apart from that issue.

**Comment by euri10** [comment]
it's been a long time I haven't played with threads but iirc you can spawn servers in threads without too much problems.
now if you want reload on top of that, that wont make it.

take a look maybe at this [closed PR](https://github.com/encode/uvicorn/pull/510) test maybe for inspiration on how to run multiple threads, hope that helps.

https://github.com/euri10/uvicorn/blob/2d08e1948c6dc6c8e7c7f7d65270595aec105ccf/tests/test_main.py#L149-L204

**Comment by euri10** [comment]
> 
> However, in the meantime...
> 
> Given that there is potentially no real reason for `sys.stdin` to be reopened in the child processes, and given that the cause of the problem is very clear, would it be at least possible to put the offending code (lines 54 and 55 in `subprocess.py`) in a `try` / `except` block so that it doesn't crash the child process? Awkwardness aside, my setup actually works really well apart from that issue.

as for this, it seems to me that 9de818ed3d617121b5b57bc9e8c508359899ff9d introduces it and it's for pdb to be able to run on a subprcess, see https://stackoverflow.com/questions/4716533/how-to-attach-debugger-to-a-python-subproccess/4717276#4717276 too
would a try / except be ok in that case ?

**Comment by likeanaxon** [comment]
@euri10 Good sleuthing! :) A few thoughts:

- Going through that SO thread, it seems to me that the effort is focused on trying to reopen `stdin` in a subclass of PDB. Is there a reason why this is this necessary to do inside `uvicorn` itself?
- Again based on the SO thread, I replaced lines 54 and 55 with the following:

```
_stdin = sys.stdin
try:
    sys.stdin = open('/dev/stdin')
    print("stdin opened successfully")
finally:
    sys.stdin = _stdin
```

Which results in the successful spawning of child processes, each of which prints `stdin opened successfully`.

- In the SO thread, they do have the attempt to reopen `stdin` in a `try/except` block. 
- It seems to me that `stdin` is reopened _**only if it is not already open**_:

```
class ForkablePdb(pdb.Pdb):

    _original_stdin_fd = sys.stdin.fileno()
    _original_stdin = None # <-- This is originally set to None...

    def __init__(self):
        pdb.Pdb.__init__(self, nosigint=True)

    def _cmdloop(self):
        current_stdin = sys.stdin
        try:
            if not self._original_stdin:
                self._original_stdin = os.fdopen(self._original_stdin_fd) # <-- ...and here it is opened only if it's still None
            sys.stdin = self._original_stdin
            self.cmdloop()
        finally:
            sys.stdin = current_stdin
```

This is the first time ever that I've had anything to do with `sys.stdin`, so I had a quick [look](https://stackoverflow.com/questions/32199552/what-is-sys-stdin-fileno-in-python) [around](https://stackoverflow.com/questions/53898231/integer-file-descriptor-0-in-open) to see what's expected. Apparently on *nix the convention is for file descriptors `0`, `1` and `2` to correspond to `stdin`, `stdout` and `stderr`, respectively. So I would have expected `sys.stdin.fileno()` to return `0`. However, at least on my system it returns an integer larger than `2` (say, `9`). So regardless of the original purpose of this part of the code, I'm not sure that `sys.stdin.fileno()` produces the intended result. To test that, I modified the original code (line `55`) from

```
sys.stdin = os.fdopen(stdin_fileno)
```

to 

```
sys.stdin = os.fdopen(0)
```

and the error was gone.

In summary, I'd say that a `try/except` is perfectly acceptable here (although I'd also argue that reopening `stdin` is not even necessary, but I'd be happy to stand corrected on that).

I would gladly create a PR for this, but before that someone with a higher level of authority should decide what the solution should look like - a `try/except`, `open('/dev/stdin')` (probably inside a `try/except`) or just remove lines `26 -- 29`, `35`,`54` and `55` altogether if those parts do not work as originally intended.

**Comment by euri10** [comment]
> In summary, I'd say that a `try/except` is perfectly acceptable here (although I'd also argue that reopening `stdin` is not even necessary, but I'd be happy to stand corrected on that).
> I would gladly create a PR for this, but before that someone with a higher level of authority should decide what the solution should look like - a `try/except`, `open('/dev/stdin')` (probably inside a `try/except`) or just remove lines `26 -- 29`, `35`,`54` and `55` altogether if those parts do not work as originally intended.

ping @blueyed, maybe you have insights on this !

the original commit seems specifically tied to pdb, which I never use and have very little experience on

on top of that it was written before we changed the default multiprocessing context to be the spawn context which is not the linux default, so maybe the interaction is different than with a plain forked process,

I'd say @likeanaxon send a PR on this the way you think it should work, then we'll advise !

**Comment by likeanaxon** [comment]
@euri10 @florimondmanca 

Hi guys,

This issue still persists in the latest release (13.4), and it is becoming increasingly unwieldy to maintain my own fork with the fix. Please let me know if there are any other merge blockers that I might not be aware of.

Cheers.

**Comment by w4tsn** [comment]
I'm also suffering from this issue while trying to start multiple uvicorn processes with `reload` set to `True`. The use-case is a main process that start's several `uvicorn.run` instances like this:

```
Process(
  target=uvicorn.run,
  kwargs={
      "app": "some.app1:main:app",
      "host": settings.UVICORN_HOST,
      "port": settings.UVICORN_PORT,
      "reload": True,
  },
).start()
... (start three more of those)
```

With two such processes it works, but it the third fails with 

```
    sys.stdin = os.fdopen(stdin_fileno)
  File "/usr/lib64/python3.10/os.py", line 1029, in fdopen
    return io.open(fd, mode, buffering, encoding, *args, **kwargs)
OSError: [Errno 9] Bad file descriptor
```

**Comment by euri10** [comment]
the use case wont likely be supported, `reload` uses multiprocess spawn under the hood so you're essentially multiprocessing^2
